<a href="https://colab.research.google.com/github/olegtyulyaev/--2/blob/main/Probability_theory_HW1_10.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Теория вероятностей 1ч**



## Домашнее задание №1

**Задача 1.**

 Из колоды в 52 карты извлекаются случайным образом 4 карты.

1.   Найти вероятность того, что все 4 карты – крести
2.   Найти вероятность, что среди 4-х карт окажется хотя бы один туз.

In [ ]:
from math import factorial
a = factorial(13) // (factorial (4) * factorial (13-4) )
b = factorial(39) // (factorial (0) * factorial (39-0) )

c = factorial(52) // (factorial (4) * factorial (52-4) )
(a*b)/c

0.0026410564225690276

In [ ]:
a = factorial(48) // (factorial (4) * factorial (48-4))
b = factorial(52) // (factorial (4) * factorial (52-4))
1-(a/b)

0.2812632745405854

**Ответ**
1.   P (4 трефы) = 0,26%
2.   P (хотя бы 1 туз) = 28,13%


**Задача 2.**

На входной двери подъезда установлен кодовый замок, содержащий десять кнопок с цифрами от 0 до 9. Код содержит три цифры, которые нужно нажать одновременно . Какова вероятность того, что человек, не знающий код, откроет дверь с первой попытки?

In [ ]:
a = factorial(10) / (factorial (3) * factorial (10-3))
a

120.0

**Ответ**

120 возможных исходов, из которых один благоприятный, искомая вероятность равна 1/120 или 0,83%

**Задача 3.**

В ящике имеется 15 деталей, из которых 9 окрашены. Рабочий случайным образом извлекает 3 детали. Какова вероятность того, что все 3 извлеченные детали окрашены?

In [ ]:
a = factorial(9) // (factorial (3) * factorial (9-3))
b = factorial(6) // (factorial (0) * factorial (6-0) )

c = factorial(15) // (factorial (3) * factorial (15-3))

(a*b)/c

0.18461538461538463

**Ответ**

P (3 из 3х окрашены) = 18,46%

**Задача 4.**

В лотерее 100 билетов. Из них 2 выигрышных. Какова вероятность того, что 2 приобретенных билета окажутся выигрышными?

In [ ]:
a = factorial(2) // (factorial (2) * factorial (2-2) )
b = factorial(98) // (factorial (0) * factorial (98-0) )

c = factorial(100) // (factorial (2) * factorial (100-2) )

(a*b)/c

0.00020202020202020202

**Ответ**

P (оба билеты выигрышные) = 0,02%


## Домашнее задание №2

**Задача 2.1**

 Вероятность того, что стрелок попадет в мишень, выстрелив один раз, равна 0.8. Стрелок выстрелил 100 раз. Найдите вероятность того, что стрелок попадет в цель ровно 85 раз.

In [ ]:
from math import factorial, exp
def bernulli(n, k, p):
    result =factorial(n)/(factorial(k)*factorial(n-k))
    return result*(p**k)*(1-p)**(n-k)

# Формула биномиального распределения: P(n)(X=k) = С(n,k) * p^k * q^(n-k) k = 85; n = 100; p = 0.8; q = 1-0.8 = 0.2

In [ ]:
print(f' P ={bernulli(100,85,0.8): .5f}')

 P = 0.04806


**Задача 2.2**

Вероятность того, что лампочка перегорит в течение первого дня эксплуатации, равна 0.0004. В жилом комплексе после ремонта в один день включили 5000 новых лампочек. Какова вероятность, что ни одна из них не перегорит в первый день? Какова вероятность, что перегорят ровно две?

In [ ]:
p = 0.0004
n = 5000
lamda = p * n
lamda

#формула Пуассона  P(m) = (lambda^m / m!) * e^(-lambda)

2.0

In [ ]:
import math
P_0 = (lamda**0/(factorial(0))) * (math.e **-lamda)
print(f' P ={(P_0): .5f}')


 P = 0.13534


In [ ]:
P_2 = (lamda**2/(factorial(2))) * (math.e **-lamda)
print(f' P ={(P_2): .5f}')


 P = 0.27067


**Ответ**

P (0 лампочек перегорит) = 13,5%

P (2 лампочки перегорят) = 27,1%

**Задача 2.3**

Монету подбросили 144 раза. Какова вероятность, что орел выпадет ровно 70 раз?

In [ ]:
print(f'P = {bernulli(144,70,0.5): .5f}')

P =  0.06281


**Ответ**

Вероятность, что орел выпадет ровно 70 раз = 6,28%

**Задача 2.4**

В первом ящике находится 10 мячей, из которых 7 - белые. Во втором ящике - 11 мячей, из которых 9 белых. Из каждого ящика вытаскивают случайным образом по два мяча.

* 2.4.1 Какова вероятность того, что все мячи белые?
* 2.4.2 Какова вероятность того, что ровно два мяча белые?
* 2.4.3 Какова вероятность того, что хотя бы один мяч белый?

In [ ]:
p4=7/10*6/9*9/11*8/10
p2=7/10*6/9*2/11*1/10+3/10*2/9*9/11*8/10+(7/10*3/9+3/10*7/9)*(9/11*2/10+2/11*9/10)
p1=1-3/10*2/9*2/11*1/10

def comb(k, n):
 return factorial(n) / (factorial(k) * factorial(n - k))

#первая и вторая корзина по 2 белых мяча
p4 = comb(2, 7) / comb(2, 10) * comb(2, 9) / comb(2, 11)

#первая или вторая корзина 2 белых мяча
p2 = (comb(2, 7) / comb(2, 10) * comb(2, 2) / comb(2, 11)) + (comb(2, 3) / comb(2, 10) * comb(2, 9) / comb(2, 11))

#первая и вторая корзина по одному белому мячу
p2 += comb(1, 7) * comb(1, 3) /  comb(2, 10) * comb(1, 9) * comb(1, 2) / comb(2, 11)

#обратная вероятность вытащить ноль белых мячей
p1 = 1 - comb(0, 7) / comb(7, 10) * comb(0, 9) / comb(9, 11)


print(f'Ответ 2.4.1 - все мячи белые = {p4:.5f}')
print(f'Ответ 2.4.2 - 2 мяча белые = {p2:.5f}')
print(f'Ответ 2.4.3 - хотя бы 1 мяч белый = {p1:.5f}')

Ответ 2.4.1 - все мячи белые = 0.30545
Ответ 2.4.2 - 2 мяча белые = 0.20485
Ответ 2.4.3 - хотя бы 1 мяч белый = 0.99985


## Домашнее задание №3

**Задача 3.1.**

Даны значения зарплат из выборки выпускников: 100, 80, 75, 77, 89, 33, 45, 25, 65, 17, 30, 24, 57, 55, 70, 75, 65, 84, 90, 150. Посчитать (желательно без использования статистических методов наподобие std, var, mean) среднее арифметическое, среднее квадратичное отклонение, смещенную и несмещенную оценки дисперсий для данной выборки.

In [ ]:
import numpy as np
from math import factorial
arr=np.array([100, 80, 75, 77, 89, 33, 45, 25, 65, 17, 30, 24, 57, 55, 70, 75, 65, 84, 90, 150])
arr

def mean_value(array):
    return sum(array)/len(array)

print(f'Среднее арифметическое М(Х) = {mean_value(arr): .2f}')

Среднее арифметическое М(Х) =  65.30


In [ ]:
def mean_square_deviation(array):
    square_dev=(array-mean_value(array))**2
    return (sum(square_dev)/len(square_dev))**(1/2)

print(f'Среднее квадратичное отклонение D(X) = {mean_square_deviation(arr): .2f}')

Среднее квадратичное отклонение D(X) =  30.82


In [ ]:
def dispers(array):
    square_dev=(array-mean_value(array))**2
    return sum(square_dev)/(len(square_dev)-1)

def dispers_slap(array):
    square_dev=(array-mean_value(array))**2
    return sum(square_dev)/(len(square_dev))

print(f'Смещенная D = {dispers(arr): .2f}')
print(f'Немещенная D = {dispers_slap(arr): .2f}')

Смещенная D =  1000.12
Немещенная D =  950.11


**Задача 3.2.**

В первом ящике находится 8 мячей, из которых 5 - белые. Во втором ящике - 12 мячей, из которых 5 белых. Из первого ящика вытаскивают случайным образом два мяча, из второго - 4 мяча. Какова вероятность того, что 3 мяча белые?

In [ ]:
import numpy as np
from math import factorial

def combinations(n,k):
    return factorial(n)/(factorial(k)*factorial(n-k))

In [ ]:
P1=(combinations(3,2)/combinations(8,2)) * (combinations(5,3) * combinations(7,1)/combinations(12,4))
print(f'Вероятность события "из первой корзины не вытащили ни одного белого мяча'
      f', из второй вытащили 3 белых мяча"\nP(A1) = {P1: .4f}'
     )

Вероятность события "из первой корзины не вытащили ни одного белого мяча, из второй вытащили 3 белых мяча"
P(A1) =  0.0152


In [ ]:
P2=(combinations(5,1)*combinations(3,1)*combinations(5,2)*combinations(7,2))/(combinations(8,2)*combinations(12,4))
print(f'Вероятность события "из первой корзины вытащили 1 белый мяч'
      f', из второй вытащили 2 белых мяча"\nP(A2) = {P2: .4f}'
     )

Вероятность события "из первой корзины вытащили 1 белый мяч, из второй вытащили 2 белых мяча"
P(A2) =  0.2273


In [ ]:
P3=(combinations(5,2)*combinations(5,1)*combinations(7,3))/(combinations(8,2)*combinations(12,4))
print(f'Вероятность события "из первой корзины вытащили 2 белых мяча'
      f', из второй вытащили 1 белый мяч"\nP(A3) = {P3: .4f}'
     )

Вероятность события "из первой корзины вытащили 2 белых мяча, из второй вытащили 1 белый мяч"
P(A3) =  0.1263


In [ ]:
print(f'Вероятность того, что из вытащенных мячей 3 белые Р(А) = {P1+P2+P3: .4f}')

Вероятность того, что из вытащенных мячей 3 белые Р(А) =  0.3687


**Задача 3.3**

На соревновании по биатлону один из трех спортсменов стреляет и попадает в мишень. Вероятность попадания для первого спортсмена равна 0.9, для второго — 0.8, для третьего — 0.6. Найти вероятность того, что выстрел произведен: a). первым спортсменом б). вторым спортсменом в). третьим спортсменом.

In [ ]:
PB=1/3
PA=PB*0.9+PB*0.8+PB*0.6
print(f'Полная вероятность наступления события А Р(А) = {PA: .4f}')

Полная вероятность наступления события А Р(А) =  0.7667


In [ ]:
PAB1=PB*0.9/PA
PAB2=PB*0.8/PA
PAB3=PB*0.6/PA
print(f'Вероятность того, что выстрел произвёл первый спортсмен: {PAB1: .4f};\n'
      f'Вероятность того, что выстрел произвёл второй спортсмен: {PAB2: .4f};\n'
      f'Вероятность того, что выстрел произвёл третий спортсмен: {PAB3: .4f}.'
     )

Вероятность того, что выстрел произвёл первый спортсмен:  0.3913;
Вероятность того, что выстрел произвёл второй спортсмен:  0.3478;
Вероятность того, что выстрел произвёл третий спортсмен:  0.2609.


**Задача 4.4**

 В университет на факультеты A и B поступило равное количество студентов, а на факультет C студентов поступило столько же, сколько на A и B вместе. Вероятность того, что студент факультета A сдаст первую сессию, равна 0.8. Для студента факультета B эта вероятность равна 0.7, а для студента факультета C - 0.9. Студент сдал первую сессию. Какова вероятность, что он учится: a). на факультете A б). на факультете B в). на факультете C?

In [ ]:
PD=0.25*0.8+0.25*0.7+0.5*0.9
print(f'Полная вероятность наступления события D, P(D) = {PD}.')

Полная вероятность наступления события D, P(D) = 0.825.


In [ ]:
PDSA=0.25*0.8/PD
PDSB=0.25*0.7/PD
PDSC=0.5*0.9/PD
print(f'Вероятность того, что студент учится на факультете А: {PDSA: .4f};\n'
      f'Вероятность того, что студент учится на факультете B: {PDSB: .4f};\n'
      f'Вероятность того, что студент учится на факультете C: {PDSC: .4f}.'
      )

Вероятность того, что студент учится на факультете А:  0.2424;
Вероятность того, что студент учится на факультете B:  0.2121;
Вероятность того, что студент учится на факультете C:  0.5455.


**Задача 3.5.**

 Устройство состоит из трех деталей. Для первой детали вероятность выйти из строя в первый месяц равна 0.1, для второй - 0.2, для третьей - 0.25. Какова вероятность того, что в первый месяц выйдут из строя: а). все детали б). только две детали в). хотя бы одна деталь г). от одной до двух деталей?

In [ ]:
P3=0.1*0.2*0.25
print(f'Вероятность того, что из строя выйдут все детали Р(3) = {P3: .4f}')

Вероятность того, что из строя выйдут все детали Р(3) =  0.0050


In [ ]:
P2=0.1*0.2*0.75+0.1*0.25*0.8+0.2*0.25*0.9
print(f'Вероятность того, что из строя выйдут только 2 детали Р(2) = {P2: .4f}')

Вероятность того, что из строя выйдут только 2 детали Р(2) =  0.0800


In [ ]:
P0=0.9*0.8*0.75
print(f'Вероятность того, что из строя не выйдет ни одной детали Р(0) = {P0: .4f}')

Вероятность того, что из строя не выйдет ни одной детали Р(0) =  0.5400


In [ ]:
P_0=1-P0
print(f'Вероятность того, что выйдет из строя хотя бы одна деталь Р(>=1) = {P_0: .4f}')

Вероятность того, что выйдет из строя хотя бы одна деталь Р(>=1) =  0.4600



## Домашнее задание №4

**Задача 4.1.**

 Случайная непрерывная величина A имеет равномерное распределение на
промежутке (200, 800].

Найдите ее среднее значение и дисперсию.

In [ ]:
def mean_and_variance(a,b):
    return f'({a};{b}]\nСреднее значение: М(А) = {(a+b)/2: .2f}\nДисперсия: D(A) = {((b-a)**2)/12: .2f}'
print(mean_and_variance(200,800))

(200;800]
Среднее значение: М(А) =  500.00
Дисперсия: D(A) =  30000.00


**Задача 4.2.**

О случайной непрерывной равномерно распределенной величине B известно, что ее
дисперсия равна 0.2.
Можно ли найти правую границу величины B и ее среднее значение зная, что левая
граница равна 0.5?
Если да, найдите ее.

In [ ]:
b=0.5+2.4**(1/2)
print(f'Левая граница распределения величины В - b = {b: .3f}\n'
      f'Среднее значение В на промежутке (0.5; {b: .3f}) M(B) = {(b+0.5)/2: .3f}'
     )

Левая граница распределения величины В - b =  2.049
Среднее значение В на промежутке (0.5;  2.049) M(B) =  1.275


**Задача 4.3.**

 Непрерывная случайная величина X распределена нормально и задана плотностью
распределения
f(x) = (1 / (4 * sqrt(2pi))) * exp((-(x+2)**2) / 32)

Найдите:

а). M(X)

б). D(X)

в). std(X) (среднее квадратичное отклонение)

In [ ]:
print('M(x) = a = -2 \n'
      'std(x) = 4 \n'
      'D(x) = std(x)^2 = 16 \n'
      )

M(x) = a = -2 
std(x) = 4 
D(x) = std(x)^2 = 16 



**Задача 4.4.**

 Рост взрослого населения города X имеет нормальное распределение, причем, средний рост равен 174 см, а среднее квадратическое отклонение равно 8 см. посчитайте, какова вероятность того, что случайным образом выбранный взрослый человек имеет рост:
1. больше 182 см?
2. больше 190 см?
3. от 166 см до 190 см?
4. от 166 см до 182 см?
5. от 158 см до 190 см?
6. не выше 150 см или не ниже 190 см?
7. не выше 150 см или не ниже 198 см?
8. ниже 166 см?
Задачу можно решить двумя способами: без использования сторонних библиотек (numpy, scipy, pandas и пр.), а затем проверить себя с помощью встроенных функций

In [ ]:
from scipy.stats import norm
from os import system

mu = 174
sigma = 8

a = 182
pa = 1-norm.cdf(a, mu, sigma)
print(f'>>> Вероятность роста больше {a} см = {pa: .4f}')

b = 190
pb = 1-norm.cdf(b, mu, sigma)
print(f'>>> Вероятность роста больше {b} см = {pb: .4f}')

v1, v2 = 166, 190
pv = norm.cdf(v2, mu, sigma)-norm.cdf(v1, mu, sigma)
print(f'>>> Вероятность роста от {v1} см до {v2} см = {pv: .4f}')

g1, g2 = 166, 182
pg = norm.cdf(g2, mu, sigma)-norm.cdf(g1, mu, sigma)
print(f'>>> Вероятность роста от {g1} см до {g2} см = {pg: .4f}')

d1, d2 = 158, 190
pd = norm.cdf(d2, mu, sigma)-norm.cdf(d1, mu, sigma)
print(f'>>> Вероятность роста от {d1} см до {d2} см = {pd: .4f}')

e1, e2 = 150, 190
pe = norm.cdf(e1, mu, sigma)+1-norm.cdf(e2, mu, sigma)
print(f'>>> Вероятность роста не выше {e1} см или не ниже {e2} см = {pe: .4f}')

yo1, yo2 = 150, 198
pyo = norm.cdf(yo1, mu, sigma)+1-norm.cdf(yo2, mu, sigma)
print(
    f'>>> Вероятность роста не выше {yo1} см или не ниже {yo2} см = {pyo: .4f}')

zh = 166
pzh = norm.cdf(zh, mu, sigma)
print(f'>>> Вероятность роста ниже {zh} см = {pzh: .4f}')

>>> Вероятность роста больше 182 см =  0.1587
>>> Вероятность роста больше 190 см =  0.0228
>>> Вероятность роста от 166 см до 190 см =  0.8186
>>> Вероятность роста от 166 см до 182 см =  0.6827
>>> Вероятность роста от 158 см до 190 см =  0.9545
>>> Вероятность роста не выше 150 см или не ниже 190 см =  0.0241
>>> Вероятность роста не выше 150 см или не ниже 198 см =  0.0027
>>> Вероятность роста ниже 166 см =  0.1587


**Задача 4.5**

 На сколько сигм (средних квадратичных отклонений) отклоняется рост человека,
равный 190 см, от
математического ожидания роста в популяции, в которой M(X) = 178 см и D(X) = 25 кв.см?

In [ ]:
Z=(190-178)/25**(1/2)
print(f'Рост человека, равный 190 см, отклоняется от математического ожидания \n'
      f' M(X) = 178 см и D(X) = 25 кв.см на Z = {Z} сигмы.')

Рост человека, равный 190 см, отклоняется от математического ожидания 
 M(X) = 178 см и D(X) = 25 кв.см на Z = 2.4 сигмы.



## Домашнее задание №5

**Задача 5.1.**

 Когда используется критерий Стьюдента, а когда Z –критерий?

**Ответ**: Критерий Стьюдента используется для определения того, существует ли значительная разница между средними значениями двух групп, обычно когда размер выборки невелик или стандартное отклонение генеральной совокупности неизвестно. Z-критерий используется, когда размер выборки велик (обычно n > 30) и известно стандартное отклонение генеральной совокупности.

**Задача 5.2.**

Утверждается, что шарики для подшипников, изготовленные
автоматическим станком, имеют средний диаметр 17 мм.
Используя односторонний критерий с α=0,05, проверить эту гипотезу, если в выборке из n=100 шариков средний диаметр оказался равным 17.5 мм, а дисперсия генеральной совокупности известна и равна 4 кв. мм.

**Решение:** Поскольку нам известны и среднее и дисперсия генеральной совокупности, то для проверки гипотезы будем использовать Z-критерий.

In [ ]:
import scipy.stats as stats

z=(17.5-17)/0.2
print(f'Наблюдаемое значение: {z: .4f}')

zt=stats.norm.ppf(0.95)
print(f'Табличное значение z-критерия: {zt: .4f}')

z<zt

Наблюдаемое значение:  2.5000
Табличное значение z-критерия:  1.6449


False

**Ответ:** нулевую гипотезу Н0 с мат  ожиданием 17 на уровне значимости 5% ***надо отвергнуть***

**Задача 5.3.**

Продавец утверждает, что средний вес пачки печенья
составляет 200 г.
Из партии извлечена выборка из 10 пачек. Вес каждой пачки составляет:
202, 203, 199, 197, 195, 201, 200, 204, 194, 190.
Известно, что их веса распределены нормально.
Верно ли утверждение продавца, если учитывать, что уровень значимости 1%? (Провести двусторонний тест.)

In [ ]:
import numpy as np
cookies = np.array([202, 203, 199, 197, 195, 201, 200, 204, 194, 190])

#Поскольку не известны ни среднее, ни дисперсия генеральной совокупности,
#то для проверки гипотезы будем использовать t-критерий.

tn=(np.mean(cookies)-200)/(np.std(cookies,ddof=1)/len(cookies)**0.5)
print(f'Наблюдаемое значение t-критерия: {tn: .4f}')


tt=stats.t.ppf(0.995,len(cookies)-1)
print(f'Табличное значение t-критерия: {tt: .4f}')

#Сравниваем t-наблюдаемое и t-табличное для определения верности Н0 гипотезы
np.abs(tn)<tt

Наблюдаемое значение t-критерия: -1.0651
Табличное значение t-критерия:  3.2498


True

**Ответ:** нулевую гипотезу Н0 с мат ожиданием 200 на уровне значимости 1% ***надо принять***

**Задача 5.4.**

Есть ли статистически значимые различия в среднем росте матерей и
дочерей?
Рост матерей 172, 177, 158, 170, 178,175, 164, 160, 169, 165
Рост взрослых дочерей: 173, 175, 162, 174, 175, 168, 155, 170, 160, 163

In [ ]:
import numpy as np
import scipy.stats as stats

mothers_height = np.array([172, 177, 158, 170, 178,175, 164, 160, 169, 165])
daughters_height = np.array([173, 175, 162, 174, 175, 168, 155, 170, 160, 163])

stats.ttest_rel(mothers_height, daughters_height)

TtestResult(statistic=0.6648478531431979, pvalue=0.5228168632983574, df=9)

**Ответ:** нулевую гипотезу Н0 ***надо принять*** статистических различий нет рост матерей и дочерей статистически не различается

# **Теория вероятностей 2ч**

## Домашнее задание №6



**Задача 6.1.**

Известно, что генеральная совокупность распределена нормально
со средним квадратическим отклонением, равным 16.
Найти доверительный интервал для оценки математического ожидания с надежностью
0.95,
если выборочная средняя M = 80, а объем выборки n = 256.

In [ ]:
import pandas as pd
import numpy as np
from statsmodels.stats.weightstats import _tconfint_generic as t_stat
from scipy.stats import norm
import math


mean_std = np.sqrt(16) / 256
print( mean_std )

left=80-1.96*16/256**(1/2)
right=80+1.96*16/256**(1/2)
print(f'95%-й доверительный интервал для оценки мат. ожидания генеральной совокупности: [{left};{right}].')



0.015625
95%-й доверительный интервал для оценки мат. ожидания генеральной совокупности: [78.04;81.96].


**Задача 6.2.**

В результате 10 независимых измерений некоторой величины X, выполненных с
одинаковой точностью,
получены опытные данные:
6.9, 6.1, 6.2, 6.8, 7.5, 6.3, 6.4, 6.9, 6.7, 6.1
Предполагая, что результаты измерений подчинены нормальному закону распределения
вероятностей,
оценить истинное значение величины X при помощи доверительного интервала,
покрывающего это
значение с доверительной вероятностью 0,95.

In [ ]:
import numpy as np
import scipy.stats as stats

#Для расчета 95%-го доверительного интервала будем использовать t-критерий

arr=np.array([6.9, 6.1, 6.2, 6.8, 7.5, 6.3, 6.4, 6.9, 6.7, 6.1])
print(f'Среднее выборочное: {np.mean(arr): .2f},\n'
      f'Размер выборки n={len(arr)},\n'
      f'СКО  по выборке(несмещенное): {np.std(arr, ddof=1): .2f}.'
     )

def t_from_table(confidens, len_array):
    alpha=(1-confidens)
    return stats.t.ppf(1-alpha/2, len_array-1)
print(f'Таб. знач-е t-критерия для 95%-го доверит. интервала выборки: {t_from_table(0.95, len(arr)): .3f}')

#расчета доверительного интервала:

def confidens_int(arr, confidens):
    return round(np.mean(arr)-t_from_table(confidens,len(arr))*np.std(arr, ddof=1)/len(arr)**0.5,3), \
           round(np.mean(arr)+t_from_table(confidens,len(arr))*np.std(arr, ddof=1)/len(arr)**0.5,3)

print(f'95%-й доверит. интервал для истин. значения Х: {confidens_int(arr, 0.95)}.')

Среднее выборочное:  6.59,
Размер выборки n=10,
СКО  по выборке(несмещенное):  0.45.
Таб. знач-е t-критерия для 95%-го доверит. интервала выборки:  2.262
95%-й доверит. интервал для истин. значения Х: (6.268, 6.912).


**Задача 6.3.**

Рост дочерей 175, 167, 154, 174, 178, 148, 160, 167, 169, 170
Рост матерей 178, 165, 165, 173, 168, 155, 160, 164, 178, 175
Используя эти данные построить 95% доверительный интервал для разности среднего
роста родителей и детей.

In [ ]:
import numpy as np

a = np.array([178, 165, 165, 173, 168, 155, 160, 164, 178, 175])
b = np.array([175, 167, 154, 174, 178, 148, 160, 167, 169, 170])


len(a)
10

len(b)
10

x_1 = np.mean(a)
print("Среднее арифметическое 2" ,x_1)

x_2 = np.mean(b)
print("Среднее арифметическое 1", x_2)

delta = x_1 - x_2
print("Разность средних", delta)

#Несмещенная дисперсия для выборки 1
D1 = np.var (a,ddof =1)
print("Несмещенная дисперия для выборки 1", D1)

#Несмещенная дисперсия для выборки 2
D2 = np.var (b,ddof =2)
print("Несмещенная дисперсия для выборки 2", D2)

#Оценка дисперсии
D = (D1 + D2)/2
print("Оценка дисперсии", D)

#Стандартная ошибка разности
SE = np.sqrt(D/10 +D/10)
print("Стандартная ошибка разности", SE)

import scipy.stats as stats
t = stats.t.ppf(0.975, 18)
print("Доверительный интервал для среднего", t)

#Находим нижнюю границу интервала
L = delta - t*SE
print("Находим нижнюю границу интервала", L )

#Находим верхнюю границу интервала
U = delta + t*SE
print("Находим нижнюю границу интервала", U)

Среднее арифметическое 2 168.1
Среднее арифметическое 1 166.2
Разность средних 1.9000000000000057
Несмещенная дисперия для выборки 1 60.10000000000001
Несмещенная дисперсия для выборки 2 102.45
Оценка дисперсии 81.275
Стандартная ошибка разности 4.031749000123892
Доверительный интервал для среднего 2.10092204024096
Находим нижнюю границу интервала -6.570390335079731
Находим нижнюю границу интервала 10.370390335079742


## Домашнее задание №7

**Задача 7.1.**

Даны две независимые выборки. Не соблюдается условие нормальности
x1 380,420, 290
y1 140,360,200,900
Сделайте вывод по результатам, полученным с помощью функции, имеются ли статистические различия между группами?

In [7]:
import numpy as np
from scipy import stats

x1 = np.array([380,420, 290])
y1 = np.array([140,360,200,900])

import numpy as np
from scipy import stats

x1 = np.array([380,420, 290])
y1 = np.array([140,360,200,900])

print(f"{stats.mannwhitneyu(x1, y1)}\np-value (0.63) > alpha (0.05) -> Нулевая гипотеза не отвергается,\nнет статистических значений между группами")

MannwhitneyuResult(statistic=8.0, pvalue=0.6285714285714286)
p-value (0.63) > alpha (0.05) -> Нулевая гипотеза не отвергается,
нет статистических значений между группами


**Задача 7.2.**

Исследовалось влияние препарата на уровень давления пациентов. Сначала
измерялось давление до приема препарата, потом через 10 минут и через 30 минут. *Есть*
ли статистически значимые различия между измерениями давления? В выборках не соблюдается условие нормальности.

1е измерение до приема препарата: 150, 160, 165, 145, 155

2е измерение через 10 минут: 140, 155, 150, 130, 135

3е измерение через 30 минут: 130, 130, 120, 130, 125

In [9]:
import numpy as np
from scipy import stats

x = np.array([150, 160, 165, 145, 155])
x10 = np.array([140, 155, 150, 130, 135])
x30 = np.array([130, 130, 120, 130, 125])

print(f"{stats.friedmanchisquare(x, x10,x30)}\np-value (0.008) < alpha (0.05) -> Нулевая гипотеза отвергается,\nа значит есть статистические различия между измерениями давления")

FriedmanchisquareResult(statistic=9.578947368421062, pvalue=0.00831683351100441)
p-value (0.008) < alpha (0.05) -> Нулевая гипотеза отвергается,
а значит есть статистические различия между измерениями давления


**Задача 7.3.**

Сравните 1 и 2 е измерения, предполагая, что 3го измерения через 30 минут не было. Есть
ли статистически значимые различия между измерениями давления?

In [11]:
import numpy as np
from scipy import stats

x = np.array([150, 160, 165, 145, 155])
x10 = np.array([140, 155, 150, 130, 135])

print(f"{stats.wilcoxon(x, x10)}\np-value (0.06) > alpha (0.05) -> Нулевая гипотеза не отвергается,\nзначит нет статистических различий между измерениями давления")

WilcoxonResult(statistic=0.0, pvalue=0.0625)
p-value (0.06) > alpha (0.05) -> Нулевая гипотеза не отвергается,
значит нет статистических различий между измерениями давления


**Задача 7.4.**

Даны 3 группы учеников плавания. Не соблюдается условие нормальности.

В 1 группе время на дистанцию 50 м составляют:
56, 60, 62, 55, 71, 67, 59, 58, 64, 67

Вторая группа : 57, 58, 69, 48, 72, 70, 68, 71, 50, 53

Третья группа: 57, 67, 49, 48, 47, 55, 66, 51, 54

Есть ли статистически значимые различия между группами?

In [14]:
x1 = np.array([56, 60, 62, 55, 71, 67, 59, 58, 64, 67])
x2 = np.array([57, 58, 69, 48, 72, 70, 68, 71, 50, 53])
x3 = np.array([57, 67, 49, 48, 47, 55, 66, 51, 54])
stats.kruskal(x1, x2, x3)

KruskalResult(statistic=5.465564058257224, pvalue=0.0650380998590494)

**Задача 7.5.**

Заявляется, что партия изготавливается со средним арифметическим 2,5 см. Проверить данную гипотезу, если известно, что размеры изделий подчинены нормальному закону распределения.
Объем выборки 10, уровень статистической значимости 5%
2.51, 2.35, 2.74, 2.56, 2.40, 2.36, 2.65, 2.7, 2.67, 2.34

In [27]:
x = np.array([2.51, 2.35, 2.74, 2.56, 2.40, 2.36, 2.65, 2.7, 2.67, 2.34])
avg = np.mean(x)
avg

2.5279999999999996

In [23]:
len(x)

10

In [24]:
np.std(x, ddof=1)

0.1572542173961923

In [25]:
t = (2.52 - 2.5)/(0.157/np.sqrt(10))
t

0.4028379184927876

In [26]:
from scipy import stats
stats.ttest_1samp(x,2.5)

TtestResult(statistic=0.5630613661802959, pvalue=0.5871439993940628, df=9)

Ответ: табличное t = 2.262, рассчётное t = 0.56. Соответственно, принимаем нулевую гипотезу.

## Домашнее задание №8

**Задача 8.1**

Даны значения величины заработной платы заемщиков банка (zp) и значения их
поведенческого кредитного скоринга (ks):
zp = [35, 45, 190, 200, 40, 70, 54, 150, 120, 110],
ks = [401, 574, 874, 919, 459, 739, 653, 902, 746, 832].

Найдите ковариацию этих двух величин с помощью элементарных действий, а затем с
помощью функции cov из numpy
Полученные значения должны быть равны.

Найдите коэффициент корреляции Пирсона с помощью ковариации и
среднеквадратичных отклонений двух признаков,
а затем с использованием функций из библиотек numpy и pandas.

In [29]:
zp = np.array([35, 45, 190, 200, 40, 70, 54, 150, 120, 110])
ks = np.array([401, 574, 874, 919, 459, 739, 653, 902, 746, 832])
cov_coef = np.mean(zp*ks) - np.mean(zp)*np.mean(ks)
cov_coef

9157.839999999997

In [30]:
np.cov(zp,ks, ddof=0)

array([[ 3494.64,  9157.84],
       [ 9157.84, 30468.89]])

In [31]:
r = cov_coef/(np.std(zp)*np.std(ks))
r

0.8874900920739158

In [32]:
np.corrcoef(zp,ks)

array([[1.        , 0.88749009],
       [0.88749009, 1.        ]])

**Задача 8.2**

Измерены значения IQ выборки студентов,
обучающихся в местных технических вузах:
131, 125, 115, 122, 131, 115, 107, 99, 125, 111.
Известно, что в генеральной совокупности IQ распределен нормально.

Найдите доверительный интервал для математического ожидания с надежностью 0.95.

In [33]:
iq = np.array([131, 125, 115, 122, 131, 115, 107, 99, 125, 111])
alfa = 0.05
t = stats.t.ppf (0.975,9)

Так как нам неизвестно среднее квадратичное отклонение генеральной совокупности, то доверительный интервал будем строить по критерию Стьюдента.



In [36]:
x_0 = np.mean(iq) - t*np.std(iq, ddof=1)/np.sqrt(len(iq))
x_0

110.55608365158724

In [37]:
x_1 = np.mean(iq) + t*np.std(iq, ddof=1)/np.sqrt(len(iq))
x_1

125.64391634841274

Ответ: Доверительный интервал [110.56; 125.64]

**Задача 8.3**

Известно, что рост футболистов в сборной распределен нормально
с дисперсией генеральной совокупности, равной 25 кв.см.

Объем выборки равен 27,
среднее выборочное составляет 174.2. Найдите доверительный интервал для
математического ожидания с надежностью 0.95.

In [41]:
sigma = 5
n = 27
avg = 174.2
alpha = 0.05

In [42]:
x_0 = avg - 1.96*sigma/np.sqrt(n)
x_0

172.31398912064722

In [43]:
x_1 = avg + 1.96*sigma/np.sqrt(n)
x_1

176.08601087935276

Ответ: Доверительный интервал [172.31; 176.09]